packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -158,58 +153,65 @@ function ackSegment(
 		op,
 		sequencedMessage: { sequenceNumber, minimumSequenceNumber },
 	} = opArgs;
+	let allowIncrementalPartialLengthsUpdate = true;
 	switch (op.type) {
 		case MergeTreeDeltaType.ANNOTATE: {
 			assert(
 				!!segment.propertyManager,
 				0x044 /* "On annotate ack, missing segment property manager!" */,
 			);
 			segment.propertyManager.ack(sequenceNumber, minimumSequenceNumber, op);
-			return true;
+			break;
 		}
 
 		case MergeTreeDeltaType.INSERT: {
 			assert(
-				segment.seq === UnassignedSequenceNumber,
+				opstampUtils.isLocal(segment.insert),
 				0x045 /* "On insert, seq number already assigned!" */,
 			);
-			segment.seq = sequenceNumber;
-			segment.localSeq = undefined;
-			return true;
-		}
 
-		case MergeTreeDeltaType.REMOVE: {
-			assertRemoved(segment);
-			segment.localRemovedSeq = undefined;
-			if (segment.removedSeq === UnassignedSequenceNumber) {
-				segment.removedSeq = sequenceNumber;
-				return true;
-			}
-			return false;
+			segment.insert = {
+				...stamp,
+				type: "insert",
+			};
+			break;
 		}
-
+		case MergeTreeDeltaType.REMOVE:
 		case MergeTreeDeltaType.OBLITERATE:
 		case MergeTreeDeltaType.OBLITERATE_SIDED: {
-			assertMoved(segment);
-			const obliterateInfo = segmentGroup.obliterateInfo;
-			assert(obliterateInfo !== undefined, 0xa40 /* must have obliterate info */);
-			segment.localMovedSeq = obliterateInfo.localSeq = undefined;
-			const seqIdx = segment.movedSeqs.indexOf(UnassignedSequenceNumber);
-			assert(seqIdx !== -1, 0x86f /* expected movedSeqs to contain unacked seq */);
-			segment.movedSeqs[seqIdx] = sequenceNumber;
-
-			if (segment.movedSeq === UnassignedSequenceNumber) {
-				segment.movedSeq = sequenceNumber;
-				return true;
-			}
+			assertRemoved(segment);
+			const latestRemove = segment.removes[segment.removes.length - 1];
+			assert(opstampUtils.isLocal(latestRemove), "Expected last remove to be unacked");
+			assert(
+				segment.removes.length === 1 ||
+					opstampUtils.isAcked(segment.removes[segment.removes.length - 2]),
+				"Expected prior remove to be acked",
+			);
 
-			return false;
+			allowIncrementalPartialLengthsUpdate = segment.removes.length === 1;
+			const removeStamp: RemoveOperationStamp = {
+				...stamp,
+				type: op.type === MergeTreeDeltaType.REMOVE ? "setRemove" : "sliceRemove",
+			};
+			segment.removes[segment.removes.length - 1] = removeStamp;
+
+			const { obliterateInfo } = segmentGroup;
+			assert(
+				(obliterateInfo !== undefined) === (op.type !== MergeTreeDeltaType.REMOVE),
+				0xa40 /* must have obliterate info */,
+			);
+			if (obliterateInfo !== undefined) {
```

Comment:
```suggestion
			if (obliterateInfo !== undefined) {
				assert(op.type !== MergeTreeDeltaType.REMOVE),
					0xa40 /* must have obliterate info */);
```

Reply:
this actually weakens the assert. Instead I'll extract things out into a couple locals to make this less weird to read.